crates/solidity/outputs/cargo/crate/src/extensions/semver/mod.rs

```diff
@@ -0,0 +1,463 @@
+pub mod parser;
+pub mod tests;
+
+use std::fmt::Display;
+
+use semver::Version;
+
+use crate::cst::NonterminalKind;
+use crate::utils::LanguageFacts;
+
+/// Parse the version pragmas in the given Solidity source code and return a list of language
+/// versions that can fulfill those requirements.
+pub fn infer_language_version(src: &str) -> Vec<Version> {
+    let parser = crate::parser::Parser::create(LanguageFacts::LATEST_VERSION).unwrap();
+    let output = parser.parse_file_contents(src);
+
+    let mut cursor = output.create_tree_cursor();
+
+    let mut found_ranges = Vec::<Range>::new();
+    while cursor.go_to_next_nonterminal_with_kind(NonterminalKind::VersionPragma) {
+        let pragma_text = cursor.node().unparse();
+        let pragma_text = pragma_text.trim();
+        let pragma_text = pragma_text
+            .strip_prefix("solidity")
+            .unwrap_or(pragma_text)
+            .trim();
+
+        found_ranges.push(parser::parse(pragma_text));
+    }
+
+    let mut matching_versions = vec![];
+    for lang_version in LanguageFacts::ALL_VERSIONS {
+        if found_ranges.iter().all(|r| r.matches(lang_version)) {
+            matching_versions.push(lang_version.clone());
+        }
+    }
+
+    matching_versions.sort();
+    matching_versions
+}
+
+/// A version that may not have specified all values. In general unspecified values will
+/// convert to `0` when resolved, however it is often important to know which values were _unspecified_
+/// (as opposed to being _specified as `0`_) when resolving ranges.
+#[derive(Copy, Clone)]
+struct PartialVersion {
+    major: VersionPart,
+    minor: VersionPart,
+    patch: VersionPart,
+}
+
+impl PartialVersion {
+    fn new() -> PartialVersion {
```

Comment:
nit suggestion: using `#derive(Default)` instead of `new()`, given that it initializes it to `None`...